hail/python/hailtop/aiotools/fs.py

```diff
@@ -632,11 +632,18 @@ async def copy_as_dir(self, sema: asyncio.Semaphore, source_report: SourceReport
             src = self.src
             if not src.endswith('/'):
                 src = src + '/'
+
+            srcentries = None
+
             try:
                 srcentries = await self.router_fs.listfiles(src, recursive=True)
-            except (NotADirectoryError, FileNotFoundError):
-                self.src_is_dir = False
-                return
+            except (NotADirectoryError, FileNotFoundError) as e:
+                if isinstance(e, FileNotFoundError):
+                    if await self.router_fs.isdir(src):
```

Comment:
If possible, I really don't want to have to do additional API requests after getting the listings.  So we should make the listfiles interface operate in a way that lets us do the right thing here without more calls.

In addition to your questions, another one is how listfiles behaves in the presence of empty directory files.  I think listfiles shouldn't throw NotADirectoryError/FileNotFoundError if the directory only has an empty directory file.